client/src/pages/Chat.tsx

```diff
@@ -69,12 +66,54 @@ export const Chat = () => {
   }, [chat, user, navigate]);
 
   useEffect(() => {
-    // scroll to bottom of the chat when getting a new msg
+    // check if the scrolling element is in viewport
+    function isInViewport(element: HTMLElement) {
+      const rect = element.getBoundingClientRect();
+      return (
+        rect.top >= 0 &&
+        (rect.bottom - 50) <= (window.innerHeight || document.documentElement.clientHeight)
+      );
+    }
     if (scrollingRef.current) {
-      (scrollingRef.current as HTMLElement).scrollIntoView({ behavior: 'smooth' });
+      if (isInViewport(scrollingRef.current)) {
+        scrollToBottom('smooth');
+        setMsgDiff(0);
+      } else {
+        // if now in viewport - add button to scroll down
+        setMsgLength(prev => {
+          const diff = (messages?.length || 0) - prev;
+          setMsgDiff(prev => prev + diff);
+          return messages?.length || 0;
+        })
+      }
     }
   }, [messages]);
```

Comment:
Refactoring for Clarity:
The logic inside the useEffect could be a bit complex, especially the part where you calculate and update the message difference (msgDiff). Consider breaking down the logic into smaller functions or adding more comments to explain each step.

Type Safety:
It's good that TypeScript types are used to ensure that element is of type HTMLElement. Make sure that the types of other variables (e.g., scrollingRef, messages, etc.) are correctly defined.

Magic Number:
The number 50 used in the isInViewport function is a magic number. Consider defining it as a constant with a meaningful name to improve code readability.

Conditional Rendering Button:
The logic to render a button for scrolling down (msgDiff) is clear, but you might want to add some visual indication or user feedback when the button is present. It could be helpful for the user to understand why the button appears.